experimental/transliterator_parser/src/compile/pass2.rs

```diff
@@ -0,0 +1,379 @@
+    ) -> Result<ds::RuleBasedTransliterator<'static>> {
+        let mut compiled_transform_groups: Vec<VarZeroVec<'static, ds::SimpleIdULE>> = Vec::new();
+        let mut compiled_conversion_groups: Vec<VarZeroVec<'static, ds::RuleULE>> = Vec::new();
+
+        for (transform_group, conversion_group) in rule_groups {
+            let compiled_transform_group: Vec<_> = transform_group
+                .into_iter()
+                .map(|id| self.compile_single_id(&id))
+                .collect();
+            compiled_transform_groups.push(VarZeroVec::from(&compiled_transform_group));
+
+            let compiled_conversion_group: Vec<_> = conversion_group
+                .into_iter()
+                .map(|rule| self.compile_conversion_rule(rule))
+                .collect();
+            compiled_conversion_groups.push(VarZeroVec::from(&compiled_conversion_group));
+        }
+
+        let res = ds::RuleBasedTransliterator {
+            visibility: true, // TODO(#3736): use metadata
+            filter: global_filter
+                .map(|f| f.code_points().clone())
+                .unwrap_or(CodePointInversionList::all()),
+            id_group_list: VarZeroVec::from(&compiled_transform_groups),
+            rule_group_list: VarZeroVec::from(&compiled_conversion_groups),
+            variable_table: self.var_table.finalize(),
+        };
+
+        Ok(res)
+    }
+
+    fn compile_conversion_rule(&mut self, rule: UniConversionRule<'p>) -> ds::Rule<'static> {
+        let ante = self.compile_section(rule.ante, parse::ElementLocation::Source);
+        let key = self.compile_section(rule.key, parse::ElementLocation::Source);
+        let post = self.compile_section(rule.post, parse::ElementLocation::Source);
+        let replacer = self.compile_section(rule.replacement, parse::ElementLocation::Target);
+        let cursor_offset = rule.cursor_offset;
+        ds::Rule {
+            ante: ante.into(),
+            key: key.into(),
+            post: post.into(),
+            replacer: replacer.into(),
+            cursor_offset,
+        }
+    }
+
+    fn compile_single_id(&mut self, id: &parse::SingleId) -> ds::SimpleId<'static> {
+        // TODO(#3736): map legacy ID to internal ID and use here
+        let id_string = format!(
+            "{}-{}{}",
+            id.basic_id.source,
+            id.basic_id.target,
+            if id.basic_id.variant.is_empty() {
+                "".to_owned()
+            } else {
+                format!("/{}", id.basic_id.variant)
+            }
+        );
+
+        ds::SimpleId {
+            id: id_string.into(),
+            filter: id
+                .filter
+                .as_ref()
+                .map(|f| f.code_points().clone())
+                .unwrap_or(CodePointInversionList::all()),
+        }
+    }
+
+    // returns the standin char
+    fn compile_variable(&mut self, var: &str) -> char {
+        if let Some(c) = self.var_to_char.get(var) {
```

Comment:
nit: prefer matching on refs over dereffing
```suggestion
        if let Some(&c) = self.var_to_char.get(var) {
```